[Yeo Zhi Shen] ip

yeozhishen opened the pull request.

yeozhishen committed:
Rename project, add greet, exit
Level 0 of ip

yeozhishen committed:
Add echo, modify gitignore
Add echo functionality to chatbot, exclude .project from being pushed to
remote

yeozhishen committed:
Add Task and TaskManager objects, refactor Fido
Abstracted the recording of tasks to the TaskManager and Task objects.
Refactored fido to have a main run function instead of having everything in main

yeozhishen committed:
Add Parser class, refactor Fido to use parser
Abstracted away the responsibility of collecting user inputs and parsing
them to the Parser class.
This avoids having to redeclare Scanner over and over again and using
the nextLine() method which may be hard to read in code

yeozhishen committed:
Added feature to mark tasks as done.
 Added getTaskIndexForMarking method in parser.
 Added handleTaskMarking and handleUnmarkingTask in Fido to not obfuscate
 code in run.
 Added methods to mark a specific task index as done/undone in
 TaskManager.
 Added methods to mark task as done/undone in Task

yeozhishen committed:
Clean code
Added extra abstraction layers. Classes such as Formatter, UserInterface
was added.
This ensures more modular code so that changes will be more predictable
and easier to trace through.

yeozhishen committed:
Fix bugs
Added exceptions to be thrown and handled when the user input is empty
ie) no spaces.
Added new error messages to be printed to the user when commands are
invalid

yeozhishen committed:
Refactor TaskManager
Added isIndexOutofRange to conform with SLAP framework

yeozhishen committed:
Clean code
Added constants for magic variables, fixed indentation, format of output
messages

yeozhishen committed:
Refactor command types
Added enumeration to map commands over to strings. This ensures that the
classes are not overloaded with magic strings

yeozhishen committed:
Add Todo, Deadline, Event
Added ability to add the above. The 3 classes above is derived from the
Task class.
Addititonal enumerators were made to account for the REGEX
expressions that have to be extracted from the input strings.
Added functionality to the parser to check for the correct events and
deadline, although there may be cases of fallthrough

yeozhishen committed:
Clean syntax
Cleaned up code syntax to ensure code readability

yeozhishen committed:
Rename enumerators
Refactored enumerators to be in line with java coding standards
(Although the enums are mainly used as constants)

Zhengwinter commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Scanner;
+public class Parser {
```
Clever use of abstraction to reduce repetition in code and improve modularization.

Zhengwinter commented on src/main/java/Fido.java:
```diff
@@ -31,15 +31,44 @@ private String processInputCommand(String command){
+        case DEADLINE:
+            return addDeadline();
+        case EVENT:
+            return addEvent();
```
Clear and descriptive naming of helper methods

Zhengwinter commented on src/main/java/ParserRegex.java:
```diff
@@ -1,14 +1,14 @@
-public enum PARSER_REGEX {
+public enum ParserRegex {
```
Good change! Classes should be PascalCase!

Zhengwinter commented on src/main/java/Fido.java:
```diff
@@ -31,15 +31,44 @@ private String processInputCommand(String command){
     }
-
+    private String addTodo() throws IllegalArgumentException {
+        if(!inputParser.isValidTodo()) {
+            throw new IllegalArgumentException("invalid Todo");
```
Apologies I may not fully understand how the throw statement works here but how do you classify an invalid statement as a invalid ToDo /Event/Deadline?

yeozhishen replied:
I defined an invalid todo when the user inputs the todo command without specifying any following tasks.
An invalid event/deadline is when the user does not specify BY or TO/FROM.

Zhengwinter replied:
Alright, so to clarify, you are accounting for the situations where the correct command words were given but the input was structured in an invalid way, which falls into the invalid toDo/Event/Deadline categories depending on the command. 
Have you also handled inputs without any command words?

yeozhishen replied:
Thanks for pointing it out! I have yet to account for empty inputs as I did think of that option yet. Going to handle it as part of this week's level 5 :)

Zhengwinter commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+        this.endTime = endTime;
+    }
+    @Override
+    public String toString() {
```
Could consider a more descriptive name instead of toString, perhaps like outputString

yeozhishen replied:
I deliberately chose to overrise the toString method so that in the even that there is a need to print the tasks/event/deadline using the System.out.print, the toString method will be invoked

Zhengwinter replied:
Ok, that makes sense. So it accounts for the cases when you print those objects (toDo/Event/Deadline) directly, so it's more comprehensive in a way?

Zhengwinter commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+        return super.toString()
+                + " (" + "from: " + startTime
+                + "to: " + endTime + ")";
+    }
```
An alternative way of structuring your Event/Deadline classes that you could consider is adding a class for handling all outputs. Then you could use some setter and getter methods for the fields of these Task subclasses. It may be cleaner and more modularized that way

Zhengwinter commented on src/main/java/Parser.java:
```diff
@@ -28,6 +28,60 @@ public String getUserInputString() {
+        return fromKeywordIndex != -1 && toKeywordIndex != -1
+                && !atEndsOfCommandArgumentList(fromKeywordIndex)
+                && !atEndsOfCommandArgumentList(toKeywordIndex)
+                && !consecutiveInArgumentList(fromKeywordIndex, toKeywordIndex);
```
Could improve readability with a switch case and a flag variable

Zhengwinter reviewed (commented):
Generally well written with clear applications of abstraction, inheritance concepts. Some minor changes have been suggested, I think the main ones include modifying the Event/Deadline classes as well as some possible readability fixes. Keep up the good work!

yeozhishen committed:
Add incorrect input handling
Modified parser to throw exceptions in the event of invalid todos,
deadlines and events, together with empty inputs.
Classification of invalid types:
Todo: no additional argument supplied after the todo keyword
Event: /to and /from were not specified or no arguments were given after
the keywords
Deadline: similar to event

Added an Enumerator for Error messages to have a fixed unified file to
specify the various error messages

yeozhishen committed:
Add additional exception checks
Added checks for todo, event and deadline, to make sure that the
commands did not include keywords from other commands

yeozhishen committed:
Merge branch 'branch-Level-5'
Merging features of handling errors onto main branch

shawnpong reviewed (commented):
I really like that you separated all your methods to improve legibility! Methods used also have good error handling (great work implementing try catch blocks).

shawnpong commented on src/main/java/Fido.java:
```diff
@@ -0,0 +1,99 @@
+        case EVENT:
+            return addEvent();
+        default:
+            return Commands.INVALID_COMMAND.string;
```
Switch case usage makes it very readable, good job!

shawnpong commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,138 @@
+                && !atEndsOfCommandArgumentList(byKeywordIndex);
+    }
+    public boolean isValidEvent() {
+        ParserRegex[] keywordsToLookOutFor = {ParserRegex.TO, ParserRegex.FROM};
```
Great idea, making sure to check for validity of the input/event!

shawnpong commented on src/main/java/Commands.java:
```diff
@@ -0,0 +1,22 @@
+    MARK("mark"),
+    UNMARK("unmark"),
+    INVALID_COMMAND("invalid command!");
+    public final String string;
```
Good way to define all the commands, to make editing your code in the future easy and more recognizable!

yeozhishen committed:
Add Index out of bounds error message
Modified unmark and mark command in fido to follow the other action
commands.
Main eror handling will be done by the Fido.run method

yeozhishen committed:
Merge branch 'branch-Level-5'

yeozhishen committed:
Add delete functionality
added delete for Level 6 and the corresponding error handling, similar
to mark/unmarking tasks

yeozhishen committed:
Add appropriate error message for delete

yeozhishen committed:
Merge branch 'master' into branch-Level-6
resolved conflicts regarding marking/unmarking tasks and error messages

yeozhishen committed:
Merge branch 'branch-Level-6'

yeozhishen committed:
Add file saving
Added ability for the application to save the tasklist in the hard disk
under the filename tasklist.
The application will also read the file upon startup and update the
tasklist with the tasks in the file

yeozhishen committed:
Add file saving feature
Added ability to save tasklist to disk, where the file is saved onto a
file named tasklist. upon startup, application will also read data from
the file and load it into memory

yeozhishen committed:
Merge branch 'branch-Level-7'
Merging working level-7 features into master branch

yeozhishen committed:
Refactor FileManager
changed file paths and directory paths to be static variables, edited
the filenames to distinguish it as a .txt file

yeozhishen committed:
Merge branch 'branch-Level-7'

yeozhishen committed:
Refactor FileManager

yeozhishen committed:
Merge branch 'branch-Level-7'

Jonoans commented on src/main/java/ErrorMessages.java:
```diff
@@ -0,0 +1,14 @@
+public enum ErrorMessages {
+    INVALID_DEADLINE("Invalid deadline. Input deadline in the form of: deadline [task] /by [time]"),
+    INVALID_EVENT("Invalid event. Input event in the form of: event [task] /from [start_time] /to [end_time]"),
+    INVALID_MARK_COMMAND("enter only 1 integer argument within range after the mark/unmark command"),
+    INVALID_TODO("Invalid todo. Input todo in the form of: todo [task]"),
+
+    MISSING_INPUT("No input detected, enter an input command"),
+    FILE_ERROR("error regarding file IO"),
+    INDEX_OUT_OF_BOUNDS("Index given is out of bounds");
+    public final String string;
+    private ErrorMessages(String string) {
+        this.string = string;
+    }
+}
```
Nice use of enums!

Jonoans commented on src/main/java/Commands.java:
```diff
@@ -0,0 +1,23 @@
+    private Commands(String string) {
+        this.string = string;
+    }
+    public static Commands getCommandEnumeration (String command) {
```
Extra space here

Jonoans commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,138 @@
+        inputString = stdin.nextLine();
+        String[] individualWords = inputString.trim().split(" ");
+        int last_word_index = individualWords.length;
+        if(individualWords.length == 0) {
```
Missing space here

Jonoans commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,36 @@
+        this.type = type;
+        setNotDone();
+    }
+   public String toString() {
```
Indentation issue here

Jonoans reviewed (commented):
Good job, you have done great in terms of avoid magic strings and numbers. Just be a little mindful of the issues with missing/extra spaces in your code.

yeozhishen committed:
Refactor code
shifted classes into their respective packages namely being utilities,
exceptions, enumerators and datastructures.
Corrected indentation errors

yeozhishen committed:
Merge branch 'branch-Level-7-refactoring'

yeozhishen committed:
Add more OOP, Refactor code
Ensured indentation is consistent and added more OOP

yeozhishen committed:
Merge pull request #1 from yeozhishen/A-MoreOOP

Add more OOP, Refactor code

yeozhishen committed:
Add find functionality
Added parser function to parse the find commands.
Added ListManager function to find based on keywords.
Edited syntax errors regarding printing of Task class.
Added correpnding errors for commands and error messages

yeozhishen committed:
Merge pull request #2 from yeozhishen/branch-Level-9

Add find functionality

yeozhishen committed:
Merge branch 'master' into branch-Level-5

yeozhishen committed:
Merge branch 'master' into branch-Level-6

yeozhishen committed:
Merge branch 'master' into branch-Level-7

yeozhishen committed:
Merge branch 'master' into A-MoreOOP

yeozhishen committed:
Add Java Documentation
Added JavaDoc comments for mainly public methods and classes for File
Manager, Formatter and Parser

yeozhishen committed:
Add More JavaDoc
Added Java Documentation for the rest of the code, fixed formatting
issues

yeozhishen committed:
Merge pull request #3 from yeozhishen/A-JavaDoc

A java doc

yeozhishen committed:
Merge branch 'master' into branch-Level-5

yeozhishen committed:
Merge branch 'master' into branch-Level-6

yeozhishen committed:
Merge branch 'master' into branch-Level-7

yeozhishen committed:
Merge branch 'master' into branch-Level-9

yeozhishen committed:
Add user guide, additional file error handling
The data file will now be recreated if the data file is corrupted
if recreation is unsuccessfull, the program will exit

yeozhishen committed:
Merge pull request #4 from yeozhishen/A-UserGuide

Add user guide, additional file error handling

yeozhishen committed:
Add Userguide Header
Made the name Fido more prominent

yeozhishen committed:
Prettify UserGuide
Fixed issues with the markdown displayed on the github pages and removed
all the additional markdown fields provided

yeozhishen committed:
Merge pull request #5 from yeozhishen/A-UserGuide

Prettify UserGuide

yeozhishen committed:
Edited Markdown Format

yeozhishen committed:
Change Markdown
Added newline between features and the feature listing itself

yeozhishen committed:
Edit Markdown
changed all the code blocks without the ` symbols with ones with. The
website fails to load them

yeozhishen committed:
Edit markdown
Add bullet spaces between bullet points so that indentation will not be
tracked

yeozhishen committed:
Change Markdown
Added Hyperlinks

yeozhishen committed:
Merge pull request #6 from yeozhishen/A-UserGuide

A user guide

yeozhishen committed:
Merge branch 'master' into branch-Level-5

yeozhishen committed:
Merge branch 'master' into branch-Level-6

yeozhishen committed:
Merge branch 'master' into branch-Level-7

yeozhishen committed:
Merge branch 'master' into branch-Level-9

yeozhishen committed:
Merge branch 'branch-Level-5' into branch-Level-9

yeozhishen committed:
Merge branch 'branch-Level-6'

yeozhishen committed:
Merge branch 'branch-Level-7'

yeozhishen committed:
Merge branch 'branch-Level-9'

yeozhishen committed:
Edit markdown
Added know issues section regarding the use of delimiters in any of the
task fields which would lead to file corruption

yeozhishen committed:
Edit markdown
Add hyperlink

yeozhishen committed:
Merge pull request #8 from yeozhishen/A-UserGuide

A user guide

yeozhishen committed:
Add more JavaDoc comments
Added additional comments for the new FileManager method and the Fido
constructor which was changed

yeozhishen committed:
Merge pull request #9 from yeozhishen/A-JavaDoc

Add more JavaDoc comments

yeozhishen committed:
Change Markdown
Added the code block for the bye command as the UI was not rendering

yeozhishen committed:
Merge pull request #10 from yeozhishen/A-UserGuide

Change Markdown

yeozhishen committed:
Add newline to ReadME
Newline was not rendering properly due to syntax issues as </br> was
used instead of <br>

yeozhishen committed:
Merge pull request #11 from yeozhishen/A-UserGuide

Add newline to ReadME

yeozhishen committed:
Fix typo in ReadMe

yeozhishen committed:
Merge pull request #12 from yeozhishen/A-UserGuide

Fix typo in ReadMe

yeozhishen committed:
Add extra asterisks in comments
Ensured that javadoc format is consistent

yeozhishen committed:
Merge pull request #13 from yeozhishen/A-JavaDoc

Add extra asterisks in comments
